### PR TITLE
crypto: Add support for conversion between different LUKS formats

### PR DESCRIPTION
**Pull request by xyakimo1**

### CODE REVIEW

**Comment by StorageGhoul** [comment]
Can one of the admins verify this patch?

**Comment by vojtechtrefny** [comment]
Jenkins, ok to test.

File: src/plugins/crypto.c [hunk 1]
```diff
@@ -2108,6 +2108,68 @@ gboolean bd_crypto_luks_set_uuid (const gchar *device, const gchar *uuid, GError
...
+                     "Failed to load device: %s", strerror_l (-ret, c_locale));
+        crypt_free (cd);
+        return FALSE;
+    }
+
+    const char *cd_type = crypt_get_type(cd);
+    if (g_strcmp0(cd_type, CRYPT_LUKS1) != 0 && \
+        g_strcmp0 (cd_type, CRYPT_LUKS2) != 0) {
+
+        g_set_error (error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_TECH_UNAVAIL,
```
Comment (vojtechtrefny) [inline src/plugins/crypto.c:2147]: Please use `BD_LVM_ERROR_FAIL` here (and in the next error msg too), `BD_LVM_ERROR_TECH_UNAVAIL` is used only in specific cases where support for something is not available due to missing or old dependencies (e.g. cryptsetup version being too old, in this case the operation is simply not possible, not unsupported).

**Review by vojtechtrefny — CHANGES_REQUESTED** [review]
Looks good to me otherwise, thank you. Also please squash the commits before we'll merge this, thanks.

File: src/lib/plugin_apis/crypto.api [hunk 2]
```diff
@@ -1076,6 +1076,21 @@ gboolean bd_crypto_luks_set_label (const gchar *device, const gchar *label, cons
  */
 gboolean bd_crypto_luks_set_uuid (const gchar *device, const gchar *uuid, GError **error);
 
+/**
+ * bd_crypto_luks_convert:
+ * @device:  a LUKS device to convert to a different version of LUKS
+ * @convert_to_luks_version: the LUKS version to convert to
```
Comment (tbzatek) [inline src/lib/plugin_apis/crypto.api:1082]: I'd suggest to use a shorter naming, i.e. `new_version` or `target_version`.

File: src/plugins/crypto.c [hunk 3]
```diff
@@ -2108,6 +2108,68 @@ gboolean bd_crypto_luks_set_uuid (const gchar *device, const gchar *uuid, GError
...
+
+    ret = crypt_load (cd, CRYPT_LUKS, NULL);
+    if (ret != 0) {
+        g_set_error (error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_DEVICE,
+                     "Failed to load device: %s", strerror_l (-ret, c_locale));
+        crypt_free (cd);
+        return FALSE;
+    }
+
+    const char *cd_type = crypt_get_type(cd);
```
Comment (tbzatek) [inline src/plugins/crypto.c:2143]: Please move all declarations at the beginning of the code block. Also, please add a space between opening parenthesis.

File: src/plugins/crypto.c [hunk 4]
```diff
@@ -2108,6 +2108,68 @@ gboolean bd_crypto_luks_set_uuid (const gchar *device, const gchar *uuid, GError
...
+    if (g_strcmp0(cd_type, CRYPT_LUKS1) != 0 && \
```
Comment (tbzatek) [inline src/plugins/crypto.c:2144]: No need for the backslash.

**Review by tbzatek — COMMENTED** [review]
Just a couple of codestyle things.

**Comment by xyakimo1** [comment]
@tbzatek, I've fixed codestyle, thank you for your notices.

File: src/plugins/crypto.c (hunk 1 above)
Comment (xyakimo1) [reply src/plugins/crypto.c]: I've changed the error to `BD_LVM_ERROR_FAIL` as you asked, but it seems to be a bit off to me -- why LVM if we are in the crypto plugin?

I didn't find explanations to different types of errors; please point me to them if they exist, and I will take them into account next time.

Also, did I specify the tech mode correctly (`BD_CRYPTO_TECH_MODE_MODIFY`)?

Comment (vojtechtrefny) [reply src/plugins/crypto.c]: > I've changed the error to BD_LVM_ERROR_FAIL as you asked, but it seems to be a bit off to me -- why LVM if we are in the crypto plugin?

Yes, using LVM error code is completely wrong here, I have no idea why I recommended this one :man_facepalming: Looking at the error codes we have in the crypto plugin, it looks like we don't have anything that would fit here so please add a new error code, probably something like `BD_CRYPTO_ERROR_CONVERT_FAILED`.

> Also, did I specify the tech mode correctly (BD_CRYPTO_TECH_MODE_MODIFY)?

Yes, that's correct.

Comment (xyakimo1) [reply src/plugins/crypto.c]: done, thank you

File: src/plugins/crypto.c (hunk 3 above)
Comment (tbzatek) [reply src/plugins/crypto.c]: One more occurrence with `convert_to_type`.

File: src/plugins/crypto.c [hunk 5]
```diff
@@ -2108,6 +2108,69 @@ gboolean bd_crypto_luks_set_uuid (const gchar *device, const gchar *uuid, GError
...
+        g_set_error (error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_DEVICE,
+                     "Failed to load device: %s", strerror_l (-ret, c_locale));
+        crypt_free (cd);
+        return FALSE;
+    }
+
+    cd_type = crypt_get_type (cd);
+    if (g_strcmp0 (cd_type, CRYPT_LUKS1) != 0 &&
+        g_strcmp0 (cd_type, CRYPT_LUKS2) != 0) {
+
```
Comment (tbzatek) [inline src/plugins/crypto.c:2147]: Extra empty line.

**Review by tbzatek — COMMENTED** [review]
Sorry for being such pedantic, it's just that we like to have a clean code.
No functional issues though, this is just about codestyle ;-)

File: src/plugins/crypto.c (hunk 5 above)
Comment (xyakimo1) [reply src/plugins/crypto.c]: Thank you for the review, I've made the fixes.

**Review by vojtechtrefny — APPROVED** [review]
Thank you

**Pull request merged**